parser-typechecker/src/Unison/Codebase/Branch.hs

```diff
@@ -211,80 +222,161 @@ branch0 terms types children edits =
   where
-    makeDeepPaths :: Map NameSegment (Branch m) -> Set Path
-    makeDeepPaths children =
-      Set.mapMonotonic Path.singleton (Map.keysSet children) <> ifoldMap go children
+    makeDeepPaths :: Branch0 m -> Set Path
+    makeDeepPaths branch = flip State.evalState mempty $ go [(mempty, branch)] mempty
       where
-        go :: NameSegment -> Branch m -> Set Path
-        go n b =
-          Set.map (Path.cons n) (deepPaths $ head b)
+        go :: [([NameSegment], Branch0 m)] -> Set Path -> DeepState m (Set Path)
+        go [] acc = pure acc
+        go ((reversePrefix, b0) : work) acc = do
+          let paths :: Set Path
+              paths =
+                if isEmpty0 b0
+                  then Set.empty
+                  else (Set.singleton . Path . Seq.fromList . reverse) reversePrefix
+          children <- deepChildrenHelper reversePrefix b0
+          go (work <> children) (paths <> acc)
 
 -- | Derive the 'deepEdits' field of a branch.
-deriveDeepEdits :: Branch0 m -> Branch0 m
+deriveDeepEdits :: forall m. Branch0 m -> Branch0 m
 deriveDeepEdits branch =
-  branch {deepEdits = makeDeepEdits (_edits branch) (nonEmptyChildren branch)}
+  branch {deepEdits = makeDeepEdits branch}
   where
-    makeDeepEdits :: Map NameSegment (EditHash, m Patch) -> Map NameSegment (Branch m) -> Map Name EditHash
-    makeDeepEdits edits children =
-      Map.mapKeysMonotonic Name.fromSegment (Map.map fst edits) <> ifoldMap go children
+    makeDeepEdits :: Branch0 m -> Map Name EditHash
+    makeDeepEdits branch = flip State.evalState mempty $ go [(mempty, branch)] mempty
       where
-        go :: NameSegment -> Branch m -> Map Name EditHash
-        go n b =
-          Map.mapKeys (Name.cons n) (deepEdits $ head b)
+        go :: [([NameSegment], Branch0 m)] -> Map Name EditHash -> DeepState m (Map Name EditHash)
+        go [] acc = pure acc
+        go ((reversePrefix, b0) : work) acc =
+          let edits :: Map Name EditHash
+              edits =
+                Map.mapKeysMonotonic
+                  (Name.fromReverseSegments . (NonEmpty.:| reversePrefix))
+                  (fst <$> _edits b0)
+           in do
+                children <- deepChildrenHelper reversePrefix b0
+                go (work <> children) (edits <> acc)
+
+type DeepState m = State (Set (NamespaceHash m), Sum Int)
+
+-- | Helper for knowing whether to descend into a child branch or not.
+-- Accepts child namespaces with previously unseen hashes, and any nested under fewer than 2 `lib` segments.
+deepChildrenHelper :: [NameSegment] -> Branch0 m -> DeepState m [([NameSegment], Branch0 m)]
+deepChildrenHelper reversePrefix b0 = do
+  case reversePrefix of
+    "lib" : _ -> modifying _2 (+ 1)
+    _ -> pure ()
+  flip
+    Monoid.foldMapM
+    (Map.toList (nonEmptyChildren b0))
+    \(ns, b) -> do
+      let h = namespaceHash b
+      result <-
+        ifM
+          (uses _1 (Set.notMember h) ||^ uses _2 (< 2))
```

Comment:
Can I suggest either:

```suggestion
          (uses (\(seenHashes, libDepth) -> Set.notMember h seenHashes || libDepth < 2))
```

Or to give names to the fields of DeepState?

Reply:
Good idea